### PR TITLE
BUG: pd.to_datetime() throws if caching is on with Null-like arguments

### LINKED ISSUE #22305: pd.to_datetime() throws if caching is on with Null-like arguments
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
result = pd.to_datetime([pd.NaT, None], cache=True)
```
#### Problem description

It results in error:

> ...
> ~/anaconda3/lib/python3.6/site-packages/pandas/core/indexes/base.py in get_indexer(self, target, method, limit, tolerance)
>    3242 
>    3243         if not self.is_unique:
> -> 3244             raise InvalidIndexError('Reindexing only valid with uniquely'
>    3245                                     ' valued Index objects')
>    3246 
> 
> InvalidIndexError: Reindexing only valid with uniquely valued Index objects
> 

#### Expected Output

The same as `result = pd.to_datetime([pd.NaT, None],cache=False)`:

`DatetimeIndex(['NaT', 'NaT'], dtype='datetime64[ns]', freq=None)`

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-53-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.2.1
pip: 10.0.1
setuptools: 36.5.0.post20170921
Cython: 0.28.3
numpy: 1.13.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 0.9.8
lxml: 3.8.0
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: 0.1.3
fastparquet: None
pandas_gbq: None
pandas_datareader: None
[paste the output of ``pd.show_versions()`` here below this line]

</details>

### PR DESCRIPTION
**Pull request by anmyachev**
- [x] closes #22305
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The problem was in contradictory behavior between `pd.Index([pd.NaT, None]).is_unique` that returns `False` and `algorithms.unique([pd.NaT, None])` that returns `array([NaT, None], dtype=object)` equivalent to the object being unique.

We are only using `Index(arg).unique()` now.

The test written by @realead was added to verify changes.

### CODE REVIEW

File: pandas/tests/series/test_timeseries.py [hunk 1]
```diff
@@ -33,6 +33,19 @@ def _simple_ts(start, end, freq='D'):
     return Series(np.random.randn(len(rng)), index=rng)
 
 
+def test_NA_values_with_cache():
+    # GH 22305
+    na_values = [None, np.nan, pd.NaT]
```
Comment (WillAyd) [inline pandas/tests/series/test_timeseries.py]: There should be a fixture for this you can use from conftest. Should also eliminate need for looping

File: pandas/core/base.py [hunk 2]
```diff
@@ -1201,7 +1205,13 @@ def _map_values(self, mapper, na_action=None):
             else:
                 values = self.values
 
-            indexer = mapper.index.get_indexer(values)
+            try:
+                indexer = mapper.index.get_indexer(values)
+            except InvalidIndexError:
```
Comment (WillAyd) [inline pandas/core/base.py]: So this is just catching all Exceptions right? Can we be more explicit?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 2 above)
Comment (gfyoung) [reply pandas/core/base.py]: @WillAyd : This is catching a specific subclass of `Exception`.  I'm not sure how this would be a catch-all.  This looks fine to me here.

File: pandas/tests/series/test_timeseries.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/series/test_timeseries.py]: @WillAyd : This is not immediately trivial.  Note that @anmyachev is using the same "fixture" twice for the loop.

I don't know off the top of my head how you can use the same fixture twice (without declaring it with two separate variables).  Any ideas?

Comment (WillAyd) [reply pandas/tests/series/test_timeseries.py]: IIRC you can alias the fixture

Comment (realead) [reply pandas/tests/series/test_timeseries.py]: In a similar situation I've used

https://github.com/realead/pandas/blob/356b8aab754c32e469508e23c447d8c650938843/pandas/conftest.py#L298

Here is an example: https://github.com/pandas-dev/pandas/blob/9caf58f3ee712a26b4ac9b9a2501141844a455ce/pandas/tests/indexes/test_base.py#L1454-L1467

Comment (anmyachev) [reply pandas/tests/series/test_timeseries.py]: Test is using `unique_nulls_fixture` now. Thanks @realead for example.

File: pandas/core/base.py [hunk 3]
```diff
@@ -1201,7 +1205,13 @@ def _map_values(self, mapper, na_action=None):
             else:
                 values = self.values
 
-            indexer = mapper.index.get_indexer(values)
+            try:
```
Comment (jreback) [inline pandas/core/base.py]: I would rather you test mapper.index.is_unique here

File: pandas/tests/series/test_timeseries.py [hunk 4]
```diff
@@ -1100,3 +1100,15 @@ def test_asarray_tz_aware(self):
             result = np.asarray(ser, dtype=object)
 
         tm.assert_numpy_array_equal(result, expected)
+
+    def test_NA_values_with_cache(self, unique_nulls_fixture,
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: this needs to be in pandas/tests/indexes/datetimes/test_tools

File: pandas/tests/series/test_timeseries.py [hunk 5]
```diff
@@ -1100,3 +1100,15 @@ def test_asarray_tz_aware(self):
...
+                                  unique_nulls_fixture2):
+        # GH 22305
+        # check pairwise, that no pair of na values
+        # is mangled
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: parametrize on cache=True/False

File: pandas/tests/series/test_timeseries.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/series/test_timeseries.py]: put near other tests for cache

File: pandas/tests/series/test_timeseries.py [hunk 6]
```diff
@@ -1100,3 +1100,15 @@ def test_asarray_tz_aware(self):
...
+        if unique_nulls_fixture is not unique_nulls_fixture2:
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: you need to test the else as well

File: pandas/tests/series/test_timeseries.py [hunk 7]
```diff
@@ -1100,3 +1100,15 @@ def test_asarray_tz_aware(self):
             result = np.asarray(ser, dtype=object)
 
         tm.assert_numpy_array_equal(result, expected)
+
+    def test_NA_values_with_cache(self, unique_nulls_fixture,
+                                  unique_nulls_fixture2):
+        # GH 22305
+        # check pairwise, that no pair of na values
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: not really sure what this comment means

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 3 above)
Comment (anmyachev) [reply pandas/core/base.py]: Outdated. @vnlitvin provided a better solution

File: pandas/tests/series/test_timeseries.py (hunk 4 above)
Comment (anmyachev) [reply pandas/tests/series/test_timeseries.py]: Done

File: pandas/tests/series/test_timeseries.py (hunk 7 above)
Comment (anmyachev) [reply pandas/tests/series/test_timeseries.py]: Removed comments for now

File: doc/source/whatsnew/v0.25.0.rst [hunk 8]
```diff
@@ -266,7 +266,7 @@ Datetimelike
 ^^^^^^^^^^^^
 
 - Bug in :func:`to_datetime` which would raise an (incorrect) ``ValueError`` when called with a date far into the future and the ``format`` argument specified instead of raising ``OutOfBoundsDatetime`` (:issue:`23830`)
--
+- Bug in :func:`to_datetime` which would raise ``InvalidIndexError`` when called with ``cache=True``, with ``arg`` including several different "not a numbers" (:issue:`22305`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: show more text of the error: InvalidIndexError: Reindexing only valid with uniquely valued Index objects

this is also confusing what you wrote: ``including several different "not a numbers"``

pls re-word

File: pandas/core/tools/datetimes.py [hunk 9]
```diff
@@ -50,11 +50,12 @@ def _maybe_cache(arg, format, cache, convert_listlike):
...
     if cache:
-        # Perform a quicker unique check
         from pandas import Index
-        if not Index(arg).is_unique:
-            unique_dates = algorithms.unique(arg)
-            cache_dates = convert_listlike(unique_dates, True, format)
+        # GH26078
+        unique_dates = Index(arg).unique()
+        if len(unique_dates) < len(arg):
+            cache_dates = convert_listlike(unique_dates.to_numpy(),
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: why did you add the ``.to_numpy()``?

remove the GH issue reference, instead restore the comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/tools/datetimes.py (hunk 9 above)
Comment (anmyachev) [reply pandas/core/tools/datetimes.py]: Before our changes `algorithms.unique(arg)` always returns `numpy.ndarray` when calling from `_maybe_cache`, so we use this cast to not change the previous behavior.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @anmyachev

**Comment by sjvdm** [comment]
This issue persists in pandas 1.2.4. Doesn't seem like this was ever merged to prod - are there perhaps any plans?
